### PR TITLE
Context menu

### PR DESCRIPTION
**Pull request by robhudson**

### CODE REVIEW

File: src/components/ContextMenu.svelte [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  function closeMenu() {
+    $showContextMenu = false;
+  }
+
+  // NOTE: I'd like to keep as much logic out of this file as much as possible
+  // and keep it focused on only displaying context menus. I'm not quite sure
+  // how to handle this with specific menu options and click event handlers,
+  // however.
```
Comment (wlach) [inline src/components/ContextMenu.svelte]: Could you use a component binding? https://svelte.dev/tutorial/component-bindings

File: src/components/explore/AggregationsOverTimeGraph.svelte [hunk 2]
```diff
@@ -54,9 +62,34 @@
...
+
+  let menuPos = { x: 0, y: 0 };
+  let zoomUrl;
+  let hov;
+
+  function generateQueryString(paramsToUpdate) {
+    const activeProductConfig = getActiveProductConfig();
+    if (!activeProductConfig) return '';
+    const params = activeProductConfig.getParamsForQueryString($store);
+    return toQueryString({ ...params, ...paramsToUpdate });
```
Comment (wlach) [inline src/components/explore/AggregationsOverTimeGraph.svelte]:
```suggestion
    return activeProductConfig ? toQueryString({ ...activeProductConfig.getParamsForQueryString($store), ...paramsToUpdate }) : '';
```

File: src/components/explore/ProbeExplorer.svelte [hunk 3]
```diff
@@ -83,13 +83,29 @@
...
+      if (found) return found;
+    }
+    return data[data.length - 1];
+  }
+
+  let ref = getDefaultReferencePoint();
+
   function setDomain(str) {
-    if (aggregationLevel === 'build_id') {
+    if (str === 'ZOOM') {
```
Comment (wlach) [inline src/components/explore/ProbeExplorer.svelte:98]: I would personally use constants for these things. e.g. `GLAM_ZOOM_KEY` or whatever

File: src/components/explore/ProbeExplorer.svelte [hunk 4]
```diff
@@ -83,13 +83,29 @@
...
+    if ($store.ref) {
+      const found = data.find((d) => d[aggregationLevel] === $store.ref);
+      if (found) return found;
+    }
+    return data[data.length - 1];
+  }
+
+  let ref = getDefaultReferencePoint();
+
   function setDomain(str) {
```
Comment (wlach) [inline src/components/explore/ProbeExplorer.svelte:97]: While you're here, could you use a more descriptive variable name?

```suggestion
  function setDomain(domain) {
```

**Review by wlach — CHANGES_REQUESTED** [review]
This looks pretty reasonable in outline-- is there any way for me to test this? See my question in #glam-dev.

**Comment by wlach** [comment]
Some additional feedback after trying this out:

* The right-click behaviour is not really discoverable. Some kind of brief text like *right click on hover to zoom* just below the graph would really help.
* It would be nice if the hovering over behaviour with the graph was disabled while the prompt (zoom from here to the ref point) is enabled
* It wasn't immediately clear to me how to reset the zoom when zoomed in. possibly you could do that by adding a "custom" setting to time horizon when zooming in, and animating the addition of that to the time horizon box: 
image

Other than that it seems to work as expected!

**Comment by wlach** [comment]
> It wasn't immediately clear to me how to reset the zoom when zoomed in. possibly you could do that by adding a "custom" setting to time horizon when zooming in, and animating the addition of that to the time horizon box:

So to be clear in that case, the time horizon box would then look like this with the custom option (which would always be selected if present):

image

In my mind, clicking on any of the other options would hide the custom zoom option.

**Comment by wlach** [comment]
BTW, I used a similar concept for mission control and I think it works pretty well: https://missioncontrol.telemetry.mozilla.org/#/release/mac/main_crashes

image

I believe GUD also does something like this.

**Comment by robhudson** [comment]
Merged some fixes to get on dev for further testing. Incremental improvements will be added in other commits.

**Pull request merged**